akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -0,0 +1,167 @@
+package akka.stream.impl.fusing
+
+import scala.collection.immutable
+import GraphInterpreter._
+
+trait In[T] {
+  def pull(): Unit
+  def cancel(): Unit
+
+  def isAvailable: Boolean
+  def getElement: T
+}
+
+trait Out[T] {
+  def push(elem: T): Unit
+  def complete(): Unit
+  def fail(ex: Throwable): Unit
+
+  def isAvailable: Boolean
+}
+
+case class GraphAssembly(
+  ins: Array[In[_]],
+  outs: Array[Out[_]]) {
+
+  def in(idx: Int): InImpl[Any] = ins(idx).asInstanceOf[InImpl[Any]]
+  def out(idx: Int): OutImpl[Any] = outs(idx).asInstanceOf[OutImpl[Any]]
+
+  def connectionCount: Int = ins.length
+}
+
+abstract class GraphStage {
+  private[stream] var ins = Vector.empty[InImpl[_]]
+  private[stream] var outs = Vector.empty[OutImpl[_]]
+
+  protected final def createIn[T](onPush: T ⇒ Unit): In[T] = {
+    val in = InImpl(this, onPush)
+    ins :+= in
+    in
+  }
+
+  protected final def createOut[T](onPull: () ⇒ Unit): Out[T] = {
+    val out = OutImpl[T](this, onPull)
+    outs :+= out
+    out
+  }
+
+}
+
+trait InterpreterContext {
+  private[stream] var interpreter: GraphInterpreter = _
+  private[stream] var id: Int = -1
+}
+
+case class InImpl[T](owner: GraphStage, onPush: T ⇒ Unit) extends In[T] with InterpreterContext {
+  override def pull(): Unit = interpreter.pull(id)
+  override def cancel(): Unit = interpreter.cancel(id)
+
+  override def getElement: T = interpreter.connectionStates(id).asInstanceOf[T]
+
+  override def isAvailable: Boolean = {
+    val state = interpreter.connectionStates(id)
+    !state.isInstanceOf[ConnectionState] && (state != null)
+  }
+}
+
+case class OutImpl[T](owner: GraphStage, onPull: () ⇒ Unit) extends Out[T] with InterpreterContext {
+  override def push(elem: T): Unit = interpreter.push(id, elem)
+  override def complete(): Unit = interpreter.complete(id)
+  override def fail(ex: Throwable): Unit = interpreter.fail(id, ex)
+
+  override def isAvailable: Boolean = interpreter.connectionStates(id) == Pushable
+}
+
+object GraphInterpreter {
+  trait ConnectionState
+  case object Pushable extends ConnectionState
+  case object Completed extends ConnectionState
+  case object Cancelled extends ConnectionState
+  case class Failed(ex: Throwable) extends ConnectionState
+
+  val NoEvent = -1
+}
+
+class GraphInterpreter(private val assembly: GraphAssembly) {
+  import GraphInterpreter._
+
+  val connectionStates = Array.fill[Any](assembly.connectionCount)(null)
+  private val mask = 255
+  private val eventQueue = Array.ofDim[Int](256)
+  private var queueHead: Int = 0
+  private var queueTail: Int = 0
+
+  def init(): Unit = {
+    var i = 0
+    while (i < assembly.connectionCount) {
+      assembly.in(i).interpreter = this
+      assembly.in(i).id = i
+      assembly.out(i).interpreter = this
+      assembly.out(i).id = i
+      i += 1
+    }
+  }
+
+  private def dequeue(): Int = {
+    if (queueHead == queueTail) -1
+    else {
+      val elem = eventQueue(queueHead & mask)
```

Comment:
well, it is an integer so there is no leak here. Can be left -1 though just to be sure.